drivers/python/rethinkdb/ast.py

```diff
@@ -698,66 +685,72 @@ def recursively_make_hashable(obj):
     return obj
 
 
-def reql_type_grouped_data_to_object(obj):
-    if 'data' not in obj:
-        raise RqlDriverError(('pseudo-type GROUPED_DATA object' +
-                              ' %s does not have the expected field "data".')
-                             % py_json.dumps(obj))
-    return dict([(recursively_make_hashable(k), v) for k, v in obj['data']])
-
-
-def reql_type_binary_to_bytes(obj):
-    if 'data' not in obj:
-        raise RqlDriverError(('pseudo-type BINARY object %s does not have ' +
-                              'the expected field "data".')
-                             % py_json.dumps(obj))
-    return RqlBinary(base64.b64decode(obj['data'].encode('utf-8')))
-
-
-def convert_pseudotype(obj, format_opts):
-    reql_type = obj.get('$reql_type$')
-    if reql_type is not None:
-        if reql_type == 'TIME':
-            time_format = format_opts.get('time_format')
-            if time_format is None or time_format == 'native':
-                # Convert to native python datetime object
-                return reql_type_time_to_datetime(obj)
-            elif time_format != 'raw':
-                raise RqlDriverError("Unknown time_format run option \"%s\"."
-                                     % time_format)
-        elif reql_type == 'GROUPED_DATA':
-            group_format = format_opts.get('group_format')
-            if group_format is None or group_format == 'native':
-                return reql_type_grouped_data_to_object(obj)
-            elif group_format != 'raw':
-                raise RqlDriverError("Unknown group_format run option \"%s\"."
-                                     % group_format)
-        elif reql_type == 'GEOMETRY':
-            # No special support for this. Just return the raw object
-            return obj
-        elif reql_type == 'BINARY':
-            binary_format = format_opts.get('binary_format')
-            if binary_format is None or binary_format == 'native':
-                return reql_type_binary_to_bytes(obj)
-            elif binary_format != 'raw':
-                raise RqlDriverError("Unknown binary_format run option \"%s\"."
-                                     % binary_format)
+class ReQLDecoder(py_json.JSONDecoder):
+    '''
+        Default JSONDecoder subclass to handle pseudo-type conversion.
+    '''
+    def __init__(self, reql_format_opts={}):
+        py_json.JSONDecoder.__init__(self, object_hook=self.convert_pseudotype)
+        self.reql_format_opts = reql_format_opts
+
+    def convert_type_time(self, obj):
```

Comment:
This should be `convert_time` right? At least it's called by that name in line 728